File: backend/src/utils/humanize.js
// backend/src/utils/humanize.js
// Utilities to simulate human-like browser behavior with Playwright

const randomInt   = (min, max) => Math.floor(Math.random() * (max - min + 1)) + min;
const randomFloat = (min, max) => Math.random() * (max - min) + min;
const sleep       = (ms) => new Promise(r => setTimeout(r, ms));

// ── Delay presets ─────────────────────────────────────────────────────────────
const humanDelay = async (speed = 'normal') => {
  const ranges = { fast: [300, 800], normal: [800, 2000], slow: [2000, 4500] };
  const [min, max] = ranges[speed] || ranges.normal;
  await sleep(randomInt(min, max));
};

// ── Type text character by character ─────────────────────────────────────────
const humanType = async (page, selector, text, options = {}) => {
  const { clear = true, speed = 'normal' } = options;
  try {
    const el = await page.waitForSelector(selector, { state: 'visible', timeout: 8000 });
    if (clear) {
      await el.click({ clickCount: 3 });
      await sleep(randomInt(50, 150));
      await page.keyboard.press('Backspace');
    }
    for (const char of String(text)) {
      await el.type(char, { delay: randomInt(35, 120) });
      if (Math.random() < 0.04) await sleep(randomInt(200, 600)); // occasional pause
    }
    if (speed !== 'fast') await sleep(randomInt(150, 400));
  } catch (err) {
    // Fallback: use fill() which is less human-like but more reliable
    try {
      const el = await page.$(selector);
      if (el) { await el.fill(String(text)); }
    } catch { /* give up */ }
  }
};

// ── Click with slight offset ──────────────────────────────────────────────────
const humanClick = async (page, selector, options = {}) => {
  const { timeout = 12000 } = options;
  try {
    const el = await page.waitForSelector(selector, { state: 'visible', timeout });
    const box = await el.boundingBox();
    if (box) {
      const x = box.x + box.width  * randomFloat(0.3, 0.7);
      const y = box.y + box.height * randomFloat(0.3, 0.7);
      await page.mouse.move(x, y, { steps: randomInt(4, 12) });
      await sleep(randomInt(40, 180));
      await page.mouse.click(x, y);
    } else {
      await el.click();
    }
    await sleep(randomInt(80, 300));
  } catch (err) {
    throw new Error(`humanClick failed on "${selector}": ${err.message}`);
  }
};

// ── Scroll gradually ─────────────────────────────────────────────────────────
const humanScroll = async (page, direction = 'down', amount = 300) => {
  const steps    = randomInt(3, 8);
  const stepSize = amount / steps;
  for (let i = 0; i < steps; i++) {
    await page.mouse.wheel(0, direction === 'down' ? stepSize : -stepSize);
    await sleep(randomInt(40, 180));
  }
};

// ── Select by label or value ──────────────────────────────────────────────────
const humanSelect = async (page, selector, value) => {
  await page.waitForSelector(selector, { state: 'visible', timeout: 8000 });
  await sleep(randomInt(100, 300));
  await page.selectOption(selector, { label: value }).catch(() =>
    page.selectOption(selector, { value }).catch(() =>
      page.selectOption(selector, { index: 1 })
    )
  );
  await sleep(randomInt(150, 400));
};

// ── Navigate with pre-delay ───────────────────────────────────────────────────
const humanNavigate = async (page, url) => {
  await sleep(randomInt(400, 1200));
  await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 30000 });
  await sleep(randomInt(900, 2000));
};

// ── Answer a field using Q&A templates ───────────────────────────────────────
// (Legacy — used by some older code paths; new code uses fillField in easy-apply.js)
const answerQuestion = async (page, fieldInfo, qaTemplates) => {
  const { label = '', type, selector } = fieldInfo;
  if (!selector) return false;
  const lower = label.toLowerCase();
  const tpl = qaTemplates.find(t => lower.includes(t.question_pattern.toLowerCase()));
  if (!tpl) return false;

  try {
    if (type === 'text' || type === 'textarea' || type === 'number' || type === 'email') {
      await humanType(page, selector, tpl.answer, { clear: true, speed: 'fast' });
    } else if (type === 'select') {
      await humanSelect(page, selector, tpl.answer);
    } else if (type === 'radio' || type === 'checkbox') {
      const opts = await page.$$(`${selector}`);
      for (const opt of opts) {
        const lbl = await opt.evaluate(el => {
          const l = el.closest('label') || document.querySelector(`label[for="${el.id}"]`);
          return l ? l.textContent.trim().toLowerCase() : '';
        });
        if (lbl.includes(tpl.answer.toLowerCase()) || tpl.answer.toLowerCase().includes(lbl)) {
          await opt.click();
          break;
        }
      }
    }
    return true;
  } catch { return false; }
};

// ── Browser launch config ─────────────────────────────────────────────────────
const getBrowserConfig = () => ({
  headless: process.env.HEADLESS === 'true',  // false by default for easier debugging
  args: [
    '--no-sandbox',
    '--disable-setuid-sandbox',
    '--disable-blink-features=AutomationControlled',
    '--disable-infobars',
    '--window-size=1440,900',
  ],
  viewport: { width: 1440, height: 900 },
});

module.exports = {
  sleep, randomInt, randomFloat,
  humanDelay, humanType, humanClick,
  humanScroll, humanSelect, humanNavigate,
  answerQuestion, getBrowserConfig,
};

File: backend/app/utils/humanize.py
from __future__ import annotations

import random
import time
from typing import Any


def random_int(minimum: int, maximum: int) -> int:
    return random.randint(minimum, maximum)


def random_float(minimum: float, maximum: float) -> float:
    return random.uniform(minimum, maximum)


def sleep(ms: int) -> None:
    time.sleep(ms / 1000)


def human_delay(speed: str = "normal") -> None:
    ranges = {
        "fast": (300, 800),
        "normal": (800, 2000),
        "slow": (2000, 4500),
    }
    minimum, maximum = ranges.get(speed, ranges["normal"])
    sleep(random_int(minimum, maximum))


def human_type(page: Any, selector: str, text: str, clear: bool = True, speed: str = "normal") -> None:
    try:
        element = page.wait_for_selector(selector, state="visible", timeout=8000)
        if clear:
            element.click(click_count=3)
            sleep(random_int(50, 150))
            page.keyboard.press("Backspace")
        for char in str(text):
            element.type(char, delay=random_int(35, 120))
            if random.random() < 0.04:
                sleep(random_int(200, 600))
        if speed != "fast":
            sleep(random_int(150, 400))
    except Exception:
        try:
            element = page.query_selector(selector)
            if element:
                element.fill(str(text))
        except Exception:
            pass


def human_click(page: Any, selector: str, timeout: int = 12000) -> None:
    try:
        element = page.wait_for_selector(selector, state="visible", timeout=timeout)
        box = element.bounding_box()
        if box:
            x = box["x"] + box["width"] * random_float(0.3, 0.7)
            y = box["y"] + box["height"] * random_float(0.3, 0.7)
            page.mouse.move(x, y, steps=random_int(4, 12))
            sleep(random_int(40, 180))
            page.mouse.click(x, y)
        else:
            element.click()
        sleep(random_int(80, 300))
    except Exception as err:
        raise RuntimeError(f'human_click failed on "{selector}": {err}') from err


def human_scroll(page: Any, direction: str = "down", amount: int = 300) -> None:
    steps = random_int(3, 8)
    step_size = amount / steps
    delta = step_size if direction == "down" else -step_size
    for _ in range(steps):
        page.mouse.wheel(0, delta)
        sleep(random_int(40, 180))


def human_select(page: Any, selector: str, value: str) -> None:
    page.wait_for_selector(selector, state="visible", timeout=8000)
    sleep(random_int(100, 300))
    try:
        page.select_option(selector, label=value)
    except Exception:
        try:
            page.select_option(selector, value=value)
        except Exception:
            page.select_option(selector, index=1)
    sleep(random_int(150, 400))


def human_navigate(page: Any, url: str) -> None:
    sleep(random_int(400, 1200))
    page.goto(url, wait_until="domcontentloaded", timeout=30000)
    sleep(random_int(900, 2000))


def answer_question(page: Any, field_info: dict[str, Any], qa_templates: list[dict[str, Any]]) -> bool:
    selector = field_info.get("selector")
    if not selector:
        return False

    label = str(field_info.get("label", "")).lower()
    template = next(
        (item for item in qa_templates if label and item.get("question_pattern", "").lower() in label),
        None,
    )
    if not template:
        return False

    field_type = field_info.get("type")
    answer = str(template.get("answer", ""))

    try:
        if field_type in {"text", "textarea", "number", "email"}:
            human_type(page, selector, answer, clear=True, speed="fast")
        elif field_type == "select":
            human_select(page, selector, answer)
        elif field_type in {"radio", "checkbox"}:
            options = page.query_selector_all(selector)
            for option in options:
                label_text = option.evaluate(
                    """(el) => {
                      const label = el.closest('label') || document.querySelector(`label[for="${el.id}"]`);
                      return label ? label.textContent.trim().toLowerCase() : '';
                    }"""
                )
                if answer.lower() in label_text or label_text in answer.lower():
                    option.click()
                    break
        return True
    except Exception:
        return False


def get_browser_config() -> dict[str, Any]:
    from ..config import HEADLESS

    return {
        "headless": HEADLESS,
        "args": [
            "--no-sandbox",
            "--disable-setuid-sandbox",
            "--disable-blink-features=AutomationControlled",
            "--disable-infobars",
            "--window-size=1440,900",
        ],
        "viewport": {"width": 1440, "height": 900},
    }

